add lease endpoint reconciler

rphillips opened the pull request:
**What this PR does / why we need it**: Adds OpenShift's LeaseEndpointReconciler to register kube-apiserver endpoints within the storage registry.

Adds a command-line argument `alpha-endpoint-reconciler-type` to the kube-apiserver.

Defaults to the old MasterCount reconciler.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes kubernetes/community#939 fixes kubernetes/kubernetes#22609

**Release note**:
```release-note
Adds a command-line argument to kube-apiserver called
--alpha-endpoint-reconciler-type=(master-count, lease, none) (default
"master-count"). The original reconciler is 'master-count'. The 'lease'
reconciler uses the storageapi and a TTL to keep alive an endpoint within the
`kube-apiserver-endpoint` storage namespace. The 'none' reconciler is a noop
reconciler that does not do anything. This is useful for self-hosted
environments.
```

/cc @lavalamp @smarterclayton @ncdc

lpabon commented:
@rphillips Could you describe which files are equal to OpenShift's and which lines/files you added/edited to make it easier to review?

ericchiang commented:
/ok-to-test

lpabon commented on pkg/master/master.go:
```diff
@@ -156,6 +179,52 @@ type completedConfig struct {
+	ttl := c.MasterEndpointReconcileTTL
+	config, err := c.StorageFactory.NewConfig(kapi.Resource("apiServerIPInfo"))
+	if err != nil {
+		glog.Fatalf("Error determining service IP ranges: %v", err)
```
Is this an acceptable pattern to exit the application at this point instead of returning an error?

rphillips replied:
Yes, I think so. It is during config setup

rphillips commented:
@lpabon The OpenShift files are marked with their original location. Everything else is written or generated for this PR

calebamiles commented:
/release-note-none

calebamiles commented:
/test pull-kubernetes-unit
/test pull-kubernetes-bazel-test
/test pull-kubernetes-node-e2e

calebamiles commented:
/release-note

klausenbusk commented on pkg/master/master.go:
```diff
@@ -155,6 +178,52 @@ type completedConfig struct {
+func (c *Config) createEndpointReconciler() EndpointReconciler {
+	glog.Infof("Using reconciler: %v", c.EndpointReconcilerType)
+	switch c.EndpointReconcilerType {
+	case "":
```
A option to disable the reconciler, would be useful for people who use self-hosted k8s.

Then they can just create a `kubernetes` service which use selector.

See: https://github.com/kubernetes/community/pull/939#issuecomment-324081192 and https://github.com/kubernetes/community/pull/939#issuecomment-326348679

rphillips replied:
i added a `none` reconciler to not do anything on the reconcile loop.

calebamiles commented:
/test pull-kubernetes-e2e-gce-bazel

lavalamp commented on pkg/election/doc.go:
```diff
@@ -0,0 +1,18 @@
+limitations under the License.
+*/
+
+// Package election provides objects for managing the list of active masters via leases.
```
Please make this a sub-directory under pkg/master/

Also, add a disclaimer here that this is not the intended way for any apiserver other than kube-apiserver to accomplish this task.

lavalamp commented on pkg/master/master.go:
```diff
@@ -85,6 +90,20 @@ const (
+
+const (
+	// DefaultMasterCountReconciler will select the original reconciler
+	MasterCountReconciler = "master-count"
```
Oh, you already have these constants! give them a type and use them everywhere. :)

lavalamp replied:
(you may need to put them in a more visible place)

smarterclayton replied:
Probably should make them CamelCase to be consistent with other constants for type

rphillips replied:
The problem was the circular dependency it caused, but I'll look into it.

lavalamp commented on pkg/master/master.go:
```diff
@@ -155,6 +180,58 @@ type completedConfig struct {
 	*Config
 }
 
+func (c *Config) createMasterCountReconciler() EndpointReconciler {
```
consider adding a "reconcilers.go" file to collect these?

lavalamp commented on pkg/master/master.go:
```diff
@@ -155,6 +180,58 @@ type completedConfig struct {
+	switch c.EndpointReconcilerType {
+	case "":
+		fallthrough
+	case MasterCountReconciler:
```
`case "", MasterCountReconciler:` is more concise.

lavalamp replied:
But consider rejecting `""` as a flag value on start up.

lavalamp commented on pkg/master/master.go:
```diff
@@ -155,6 +180,58 @@ type completedConfig struct {
+	case NoneEndpointReconciler:
+		return c.createNoneReconciler()
+	default:
+		glog.Fatalf("Reconciler not implemented: %v", c.EndpointReconcilerType)
```
%q

lavalamp commented on pkg/master/master.go:
```diff
@@ -155,6 +180,58 @@ type completedConfig struct {
+		StorageConfig:           endpointConfig,
+		Decorator:               generic.UndecoratedStorage,
+		DeleteCollectionWorkers: 0,
+		ResourcePrefix:          c.StorageFactory.ResourcePrefix(kapi.Resource("endpoints")),
```
I would use something that clearly won't collide with the endpoints api. `"kube-apiserver-endpoint"`?

smarterclayton replied:
@ncdc I don't remember whether there was a reason we didn't do that before?

lavalamp reviewed (commented):
I haven't yet reviewed the code that you're upstreaming from openshift.

smarterclayton commented:
@ncdc since you're the original author if you want to review

bgrant0607 commented:
@rphillips Was there a specific aspect you thought I should look at?

This just changes how Endpoints is populated for the apiserver?

This needs a better release note. See https://github.com/kubernetes/community/issues/484 for guidance.

rphillips commented:
@lavalamp latest commits addresses all your comments I believe. In addition to changing the command-line argument to read:

```
     --alpha-endpoint-reconciler-type string                   Use an endpoint reconciler (master-count, lease, none) (default "master-count")
```

ncdc commented:
I'm afk today. I'll try to look after Labor Day when I'm back working.

On Thu, Aug 31, 2017 at 9:30 PM Clayton Coleman <notifications@github.com>
wrote:

> @ncdc <https://github.com/ncdc> since you're the original author if you
> want to review
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/51698#issuecomment-326463258>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAABYmb2YYpgz03aG8IOsqi5Q9JUY4ITks5sd147gaJpZM4PIsxm>
> .
>

euank commented:
/test pull-kubernetes-kubemark-e2e-gce

rphillips commented:
@ncdc just a friendly reminder when you get back to check this PR out.

ncdc commented on pkg/master/master.go:
```diff
@@ -85,6 +90,10 @@ const (
 	DefaultEndpointReconcilerInterval = 10 * time.Second
+	// DefaultEndpointReconcilerTTL is the default TTL timeout for the storage layer
+	DefaultEndpointReconcilerTTL = 15
+	// DefaultStorageEndpoint is the default storage endpoint endpoint for the lease controller
```
endpoint endpoint

ncdc commented on pkg/master/master.go:
```diff
@@ -135,12 +144,18 @@ type Config struct {
 	// same value for this field. (Numbers > 1 currently untested.)
 	MasterCount int
+
+	// out of the kubernetes service record. It is not recommended to set this value below 15s.
```
Missing beginning of comment?

ncdc commented on pkg/master/master.go:
```diff
@@ -135,12 +144,18 @@ type Config struct {
 	MasterCount int
+
+	// out of the kubernetes service record. It is not recommended to set this value below 15s.
+	MasterEndpointReconcileTTL int
```
Make this a time.Duration?

ncdc commented on pkg/master/master.go:
```diff
@@ -191,10 +257,12 @@ func (c *Config) Complete() completedConfig {
+	}
+
 	if c.EndpointReconcilerConfig.Reconciler == nil {
-		// use a default endpoint reconciler if nothing is set
```
keep the comment?

ncdc commented on pkg/master/reconcilers/doc.go:
```diff
@@ -0,0 +1,21 @@
+// Package reconcilers provides objects for managing the list of active masters.
+// NOTE: The Lease reconciler is not the intended way for any apiserver other
+// than kube-apiserver to accomplish the task of Endpoint registration. This is
+// a special case fo the time being.
```
fo -> for

ncdc commented on pkg/master/reconcilers/lease.go:
```diff
@@ -0,0 +1,251 @@
+
+/*
+Original Source:
+https://github.com/openshift/origin/blob/master/pkg/cmd/server/election/lease_endpoint_reconciler.go
```
I'd recommend replacing "master" with a real commit so it's a permalink.

rphillips commented:
@ncdc fixed those up in commit df748fd

I researched changing `MasterEndpointReconcileTTL` into a duration. The underlying storageapi requires a uint64 for the TTL. I believe the conversion from a duration to an uint64 does not make sense in this case, so I have added a comment that it is in seconds.

ncdc commented:
@rphillips `time.Duration` is a typedef'd `int64`. Not quite a `uint64` but close enough. I think it makes more sense to have this as a `time.Duration` and just do the conversion on the way out, don't you think?

rphillips commented:
@ncdc didn't realize that... I'll make the change

rphillips commented:
fixed in 62d05a3

rphillips commented:
added a release note to the initial note

euank commented on pkg/master/reconcilers/none.go:
```diff
@@ -0,0 +1,38 @@
+)
+
+// NoneEndpointReconciler allows for the endpoint reconciler to be disabled
+type NoneEndpointReconciler struct{}
```
Is there any reason this type is exported?

Same goes for other reconciler structs.

ncdc replied:
I know that OpenShift often configures and starts up things like the master, and it needs full access to the types. cc @deads2k @liggitt

euank replied:
Why though? At least for `None` it can certainly still call `NewNoneEndpointReconciler`. The type adds nothing here since the constructor is still exported.

ncdc replied:
Sure, if there's a constructor, can just export the interface & the constructors.

euank commented on pkg/master/reconcilers/reconcilers.go:
```diff
@@ -0,0 +1,55 @@
+
+const (
+	// MasterCountReconcilerType will select the original reconciler
+	MasterCountReconcilerType = "master-count"
```
Let's make a wrapper type for this since these are the only values. That'll make it clearer than 'string' when referencing the type.

Specifically something like:

```go
type ReconcilerType string

const (
   MasterCountReconcilerType ReconcilerType = "master-count"
   LeaseEndpointReconcilerType = "lease"
   NoneEndpointReconcilerType = "none"
)
```

And from there, we can also change the type name to just `Type` since it's already in a `reconciler` package; `reconcilers.Type` vs `reconcilers.ReconcilerType`.

rphillips replied:
I think it adds too much boilerplate to convert from the named type to a string (ie, for the strings.Join() call) and elsewhere in the code. The code would need to be instrumented to cast from the type to a string and vice versa.

Golang 1.9 has aliases, perhaps it could be tweaked in the future to be a golang type alias?

euank replied:
Type aliases aren't meant for this iiuc, so I don't think that helps.

For the sake of `AllTypes`, I think using a `[]string` for that seems fine. I can't think of other places, besides the flags, where there should be any boilerplate?

Alternately, you could make `[]ReconcilerType` have a `.String()` method which joins them by commas and have the flags print `AllTypes.String()`

rphillips replied:
Fixed, though I named the function 'Names' instead of forcing a format

euank commented on pkg/master/reconcilers/none.go:
```diff
@@ -0,0 +1,38 @@
+
+// NewNoneEndpointReconciler creates a new EndpointReconciler that reconciles based on a
+// nothing. It is a noop.
+func NewNoneEndpointReconciler() *NoneEndpointReconciler {
```
I'd prefer if this return the `EndpointReconciler` interface instead.

euank commented on pkg/master/reconcilers/lease.go:
```diff
@@ -0,0 +1,252 @@
+package reconcilers
+
+/*
+Original Source:
```
I don't think this comment helps much. If people are `git blaming`, they'll find the diff where it was created / moved anyways.

ncdc replied:
It's from OpenShift, not Kubernetes though.

euank replied:
Oops, missed that! Yeah, I'm happy with leaving it then.

euank reviewed (changes requested):
I left a couple type ergonomic complaints. I didn't closely review the actual code functionality since I'm not all that familiar with it.

rphillips commented:
fixed up the exports on the reconcilers and returning the EndpointReconciler interface.

rphillips commented:
pushed a tweak to make those types really types.

euank commented on cmd/kube-apiserver/app/options/options.go:
```diff
@@ -70,6 +71,9 @@ type ServerRunOptions struct {
 	EnableAggregatorRouting bool
+
+	MasterCount            int
+	EndpointReconcilerType string
```
shouldn't this be a `reconcilers.Type`?

rphillips replied:
That is the string from the commandline

euank commented on cmd/kube-apiserver/app/options/options.go:
```diff
@@ -164,6 +169,9 @@ func (s *ServerRunOptions) AddFlags(fs *pflag.FlagSet) {
 	fs.IntVar(&s.MasterCount, "apiserver-count", s.MasterCount,
 		"The number of apiservers running in the cluster, must be a positive number.")
 
+	fs.StringVar(&s.EndpointReconcilerType, "alpha-endpoint-reconciler-type", string(s.EndpointReconcilerType),
```
I think it would be slightly more elegant to implement `pflag.Value` on `reconcilers.Type` and use `fs.Var`.

`s.ServiceNodePortRange` already does this as a precedent.

liggitt replied:
that seems like overkill... there's more of a case to be made for a port range flag type than a single-use typedef

euank replied:
Yeah, you're probably right.

I'm fine leaving it as is; I acknowledge this has hit bikeshedding levels.

euank commented on pkg/master/reconcilers/mastercount.go:
```diff
@@ -0,0 +1,204 @@
+		}
+	}
+	for _, addr := range sub.Addresses {
+		if addr.IP == ip {
```
Iirc, this technically should be `addr.IP.Equal(ip)` since an ipv4 address, in go, can be represented by either a 4 byte or 16 byte slice, and thus two ipv4 addresses might be `.Equal` but not `==`.

rphillips replied:
in this case it is stored as a string, and not as a net.IP object

euank reviewed (commented):
I've nitted a little more about how it could be a little cleaner.

After seeing it with the stuttering removed, it also seems like `reconciler` vs `reconcilers` reads better, e.g. `reconciler.NewNoneEndpointReconciler()` vs `reconsilers.` 

I know my review's basically just bikeshedding at this point, sorry!

lpabon commented on pkg/master/reconcilers/reconcilers.go:
```diff
@@ -38,18 +38,32 @@ type EndpointReconciler interface {
 }
 
+// Type the reconciler type
+type Type string
```
You may need to be more descriptive, like: `ReconcilerType`

euank replied:
We ended up here because before it was stuttering (see "Avoid Stuttering" [here](https://blog.golang.org/package-names)).

smarterclayton commented:
Api changes are fine

/approve

Commit:
add lease endpoint reconciler

fixes kubernetes/community#939
fixes kubernetes/kubernetes#22609

rphillips commented:
rebased

calebamiles commented:
/test pull-kubernetes-federation-e2e-gce

smarterclayton commented:
/lgtm
/retest

But we need to @kubernetes/sig-release-feature-requests to approve this if the intent is still to deliver for 1.8.

alexbrand commented:
Great to see this. Thanks!

I think user-facing documentation about the problem and the reconcilers would be useful? Maybe in https://kubernetes.io/docs/admin/high-availability/?

Happy to open a PR if it makes sense.

luxas commented:
@alexbrand An user-facing docs PR for this would make sense to me :+1: 

@rphillips Please graduate this to beta in v1.10 and add e2e tests https://github.com/kubernetes/kubernetes/issues/57617

tengqm commented:
@luxas FYI, this has been documented kubernetes/website#6695.

luxas commented:
Great! Thanks @tengqm!

rphillips commented:
@tengqm I've been on vacation. thanks!
